proposals/jvm-expose-boxed.md

```diff
@@ -0,0 +1,268 @@
+
+**Compatibility with current compilation scheme**: if we alter the way inline classes are currently compiled in a non-backward compatible way, we could create a huge split in the community. Or even worse, we will end up supporting two different schemes in the compiler.
+
+## Expose boxed constructors
+
+The [current compilation scheme](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md#inline-classes-abi-jvm) exposes the constructor of the boxed class as _synthetic_, which makes it unavailable from Java. This constructor does _not_ execute the `init` block, it just boxes the value.
+
+Note that when boxing needs to occur, the compilation scheme does not call this constructor directly. Rather, it uses the static `box-impl` method for that type.
+
+```kotlin
+fun positiveIntSingleton(number: PositiveInt): List<PositiveInt> = listOf(number)
+// is "translated" into the following code
+fun positiveIntSingleton-3bd7(number: Int): List<PositiveInt> = listOf(PositiveInt.box-impl(number))
+```
+
+We propose a new compilation scheme which _replaces_ the previous one with respect to constructors.
+
+- The previous synthetic constructor, which does not execute the `init` block, now takes an additional `BoxingConstructorMarker` argument. This argument is always passed as `null`, but allows us to differentiate the constructor from other (like `DefaultConstructorMarker`). The implementation of `box-impl` is updated to call this constructor.
+- We get a new constructor with the same visibility in the constructor as the one defined in the value class. This constructor is exposed to Java, and should execute any `init` block; in the current compilation scheme, this amounts to calling `constructor-impl` over the input value.
+
+```kotlin
+@JvmInline value class PositiveInt(val number: Int) {
+  init { require(number >= 0) }
+}
+
+// compiles down to
+class PositiveInt {
+  private synthetic <init>(number: Int, marker: BoxingConstructorMarker)
+
+  public <init>(number: Int): void = <init>(constructor-impl(number), null)
+
+  public static constructor-impl(Int): Int  // executes the 'init' block
+  public static box-impl(number: Int): PositiveInt = <init>(number, null)
+  // and others
+}
+```
+
+Note that this change is binary compatible with previous users of this code:
+
+- The previous compilation scheme never calls the constructor directly; it always uses `box-impl` if boxing is required in the Kotlin side. But the signature of `box-impl` is not changed, only its implementation.
+- Java users were not able to call the constructor, since it was marked as synthetic (and private). So exposing the constructor is compatible.
+- In the scenario in which the constructor was called directly (for example, by reflection), the code is still correct. There might be some performance hit, since now the `init` block is executed, but this is not a big concern.
```

Comment:
Maybe I misunderstand this, but I don't see how the code calling the constructor directly by reflection is still working, because now that constructor has an extra parameter.